feat(dev-infra): migrate to unified commit message types in commit message linting

josephperrott opened the pull request:
Previously commit message types were provided as part of the ng-dev config in the repository
using the ng-dev toolset.  This change removes this configuration expectation and instead
predefines the valid types for commit messages.

Additionally, with this new unified set of types requirements around providing a scope have
been put in place.  Scopes are either required, optional or forbidden for a given commit
type.

josephperrott committed:
feat(dev-infra): migrate to unified commit message types in commit message linting

Previously commit message types were provided as part of the ng-dev config in the repository
using the ng-dev toolset.  This change removes this configuration expectation and instead
predefines the valid types for commit messages.

Additionally, with this new unified set of types requirements around providing a scope have
been put in place.  Scopes are either required, optional or forbidden for a given commit
type.

devversion commented on dev-infra/commit-message/config.ts:
```diff
@@ -30,3 +29,56 @@ export function getCommitMessageConfig() {
 }
+
+/** Scope requirement level to be set for each commit type.  */
+export enum SCOPE_REQUIREMENT {
```
Nit: I'd not use upper case for the enum name, but rather for the values. That would be more aligned with what we do in the whole code base (and the [coding style](https://google.github.io/styleguide/jsguide.html#naming-enum-names))

devversion commented on dev-infra/commit-message/validate.ts:
```diff
@@ -86,11 +86,33 @@ export function validateCommitMessage(
+  const scopeRequirementForType = COMMIT_TYPES[commit.type].scope;
+
+  if (scopeRequirementForType === SCOPE_REQUIREMENT.Forbidden) {
+    if (commit.scope) {
```
You could combine that into a single `if` with less nesting (same below).

devversion reviewed (approved):
Nice! One related question. Do we know how that can work in the future with Angular Components? e.g. we would want to enforce that scopes have a specific format. e.g. `packageName/compName`. I assume we could add configuration/checks for this too in the future, right?

devversion commented:
@josephperrott Also, do you plan on removing the `types` in the `.ng-dev` configuration or will that be a follow-up?

gkalpak commented on dev-infra/commit-message/config.ts:
```diff
@@ -30,3 +29,56 @@ export function getCommitMessageConfig() {
+/** The valid commit types for Angular commit messages. */
+export const COMMIT_TYPES: {[key: string]: CommitType} = {
+  build: {
+    value: 'build',
```
OOC, what is the benefit of having the `value` property here? It doesn't seem to be used and it adds more overhead to keep it in sync with the key.

josephperrott replied:
I pulled this change out of a large change, and it's an artifact from that.  I removed it for now and we can add it in later in a another change.

gkalpak reviewed (approved):
One nit. Otherwise lgtm :+1:

josephperrott committed:
fixup! feat(dev-infra): migrate to unified commit message types in commit message linting

josephperrott committed:
build: update ng-dev config file for new commit message configuration

Removes the commit message types from the config as they are now staticly
defined in the dev-infra code.

josephperrott commented:
@devversion added a commit with an update to the commit message config.
